## DNN 모델

#### 라이브러리 임포트

In [13]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

#### 데이터 불러오기

In [14]:
df = pd.read_csv('pd.csv')

print(df.head())

   id  time    AMT  rate    wt     ht  age  sex   bis
0   1   0.0    0.0  0.00  81.0  173.0   33    1  97.6
1   1  13.1  164.0  0.33  81.0  173.0   33    1  97.7
2   1  13.5  164.0  0.33  81.0  173.0   33    1  97.7
3   1  13.7  164.0  0.33  81.0  173.0   33    1  97.7
4   1  13.8  164.0  0.33  81.0  173.0   33    1  97.7


#### 데이터 전처리

In [17]:
X = df.drop('bis', axis=1)  # 'bis' 열을 제외한 나머지를 입력 변수로 사용
y = df['bis']  # 'bis' 열을 타겟 변수로 사용

# 훈련 데이터와 테스트 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [18]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape) 

(3164, 8)
(791, 8)
(3164,)
(791,)


In [19]:
# 데이터 표준화
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

#### 모델 생성

In [20]:
model = Sequential()
model.add(Dense(128, activation='relu', input_shape=(X_train_scaled.shape[1],)))
model.add(Dropout(0.2)) 
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.2))  
model.add(Dense(32, activation='relu'))
model.add(Dense(1))  # 회귀 문제
model.compile(optimizer='adam', loss='mean_squared_error')

#### 모델 훈련

In [21]:
early_stopping = tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)

model.fit(X_train_scaled, y_train, epochs=100, batch_size=32, 
          validation_split=0.2, callbacks=[early_stopping])


Epoch 1/100
80/80 [==============================] - 1s 4ms/step - loss: 2780.0581 - val_loss: 549.6769
Epoch 2/100
80/80 [==============================] - 0s 3ms/step - loss: 475.1015 - val_loss: 357.5273
Epoch 3/100
80/80 [==============================] - 0s 3ms/step - loss: 410.5110 - val_loss: 324.0281
Epoch 4/100
80/80 [==============================] - 0s 3ms/step - loss: 385.5336 - val_loss: 309.5491
Epoch 5/100
80/80 [==============================] - 0s 3ms/step - loss: 362.2740 - val_loss: 298.2649
Epoch 6/100
80/80 [==============================] - 0s 3ms/step - loss: 352.5719 - val_loss: 290.7079
Epoch 7/100
80/80 [==============================] - 0s 3ms/step - loss: 346.8583 - val_loss: 290.8057
Epoch 8/100
80/80 [==============================] - 0s 3ms/step - loss: 346.2339 - val_loss: 282.6072
Epoch 9/100
80/80 [==============================] - 0s 2ms/step - loss: 338.8705 - val_loss: 281.0932
Epoch 10/100
80/80 [==============================] - 0s 2ms/step - loss

#### 모델 평가

In [22]:
model.evaluate(X_test_scaled, y_test)

25/25 [==============================] - 0s 1ms/step - loss: 201.0011


201.00108337402344

In [25]:
y_pred = model.predict(X_test_scaled).flatten()

# MSE 계산
mse = mean_squared_error(y_test, y_pred)
print("MSE", mse)

# MAE 계산
mae = mean_absolute_error(y_test, y_pred)
print("MAE", mae)

# MRAE 계산
mrae = np.mean(np.abs((y_test - y_pred) / y_test))
print("MRAE", mrae)


25/25 [==============================] - 0s 1000us/step
MSE 201.00113083229311
MAE 11.223787843533023
MRAE 0.20415082092718403


## MeNet 모델

In [43]:
pd.set_option('display.max_rows', 350)
print(df.shape)

print(df.head(350))

(3955, 9)
     id   time      AMT       rate    wt     ht  age  sex   bis
0     1    0.0    0.000   0.000000  81.0  173.0   33    1  97.6
1     1   13.1  164.000   0.330000  81.0  173.0   33    1  97.7
2     1   13.5  164.000   0.330000  81.0  173.0   33    1  97.7
3     1   13.7  164.000   0.330000  81.0  173.0   33    1  97.7
4     1   13.8  164.000   0.330000  81.0  173.0   33    1  97.7
5     1   14.0  164.000   0.330000  81.0  173.0   33    1  95.0
6     1   14.2  164.000   0.330000  81.0  173.0   33    1  89.9
7     1   14.3  164.000   0.330000  81.0  173.0   33    1  57.0
8     1   14.5  164.000   0.330000  81.0  173.0   33    1  51.3
9     1   14.7  164.000   0.330000  81.0  173.0   33    1  52.1
10    1   14.8  164.000   0.330000  81.0  173.0   33    1  44.8
11    1   15.0  164.000   0.330000  81.0  173.0   33    1  45.1
12    1   15.2  164.000   0.330000  81.0  173.0   33    1  47.1
13    1   15.3  164.000   0.330000  81.0  173.0   33    1  42.7
14    1   15.5  164.000   0.33

In [44]:
unique_id_count = df['id'].nunique()

print("id의 개수:", unique_id_count)

id의 개수: 31


In [45]:
id_counts = df.groupby('id').size()

print(id_counts)

id
1     128
2     133
3     150
4     137
5     128
6     139
7     133
8     123
9     124
10    120
11    117
12    132
13    119
14    139
15    122
16    119
17    126
18    123
19    126
20    133
21    120
22    118
23    135
24    119
25    133
26    132
27    126
28    131
29    120
30    132
31    118
dtype: int64


In [46]:
average_count = id_counts.mean()

print("각 'id'당 평균 행 개수:", average_count)

각 'id'당 평균 행 개수: 127.58064516129032


#### 데이터 정보

1. (3955, 9)
2. 31명의 환자에 대한 장기적인 정보
3. 한 환자당 127개의 데이터

약리동태학(Pharmacodynamics) 데이터(pd.csv)

- 고정 효과와 무작위 효과를 모두 고려하는 Mixed Effects Model 

#### 라이브러리 임포트

In [ ]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import mixedlm

#### 데이터 전처리

In [48]:
# Mixed Effects Model 정의 및 적합
# 'bis'를 종속 변수로, 'AMT', 'Rate', 'WT', 'HT', 'Age', 'Sex'를 고정 효과로 사용
# 'id'를 무작위 효과로 사용
model = mixedlm("bis ~ AMT + rate + wt + ht + age + C(sex)", df, groups=df["id"])
result = model.fit()

# 결과 출력
print(result.summary())

          Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: bis        
No. Observations: 3955    Method:             REML       
No. Groups:       31      Scale:              487.9336   
Min. group size:  117     Log-Likelihood:     -17881.1356
Max. group size:  150     Converged:          Yes        
Mean group size:  127.6                                  
---------------------------------------------------------
              Coef.  Std.Err.   z    P>|z|  [0.025 0.975]
---------------------------------------------------------
Intercept     39.817   29.594  1.345 0.178 -18.186 97.819
C(sex)[T.1]   -5.649    2.920 -1.935 0.053 -11.372  0.073
AMT           -0.032    0.007 -4.627 0.000  -0.046 -0.019
rate          -0.149    0.097 -1.539 0.124  -0.339  0.041
wt            -0.007    0.130 -0.052 0.959  -0.261  0.247
ht             0.168    0.206  0.816 0.414  -0.235  0.572
age            0.081    0.050  1.642 0.101  -0.016  0.179
Group Var     13.868    

In [49]:
predicted_values = result.predict()

actual_values = df["bis"]

mse = ((predicted_values - actual_values) ** 2).mean()
print("MSE:", mse)

mae = abs(predicted_values - actual_values).mean()
print("MAE:", mae)

mrae = (abs(predicted_values - actual_values) / actual_values).mean()
print("MRAE:", mrae)

MSE: 498.770647299748
MAE: 19.469705556024948
MRAE: 0.361339405034508


선형 모델을 사용하여 모델 성능이 좋지 않음. 
비선형 모델을 사용하여, Mixed Effects Model에 DNN을 적용하는 방법을 사용해 볼 필요가 있다. 